tools/clang/lib/Sema/SemaDXR.cpp

```diff
@@ -1138,171 +1138,198 @@ void DiagnoseRaytracingPayloadAccess(clang::Sema &S,
   visitor.diagnose(TU);
 }
 
-void DiagnoseEntry(Sema &S, FunctionDecl *FD) {
-  auto Attr = FD->getAttr<HLSLShaderAttr>();
-  if (!Attr)
-    return;
-
-  DXIL::ShaderKind Stage = ShaderModel::KindFromFullName(Attr->getStage());
-  switch (Stage) {
-  case DXIL::ShaderKind::Pixel:
-  case DXIL::ShaderKind::Vertex:
-  case DXIL::ShaderKind::Geometry:
-  case DXIL::ShaderKind::Hull:
-  case DXIL::ShaderKind::Domain:
-  case DXIL::ShaderKind::Compute:
-  case DXIL::ShaderKind::Library:
-  case DXIL::ShaderKind::Mesh:
-  case DXIL::ShaderKind::Amplification:
-  case DXIL::ShaderKind::Invalid:
-    return;
+void DiagnoseCallableEntry(Sema &S, FunctionDecl *FD, HLSLShaderAttr *Attr) {
+  if (!FD->getReturnType()->isVoidType())
+    S.Diag(FD->getLocation(), diag::err_raytracing_must_return_void);
+
+  if (FD->getNumParams() != 1)
+    S.Diag(FD->getLocation(), diag::err_raytracing_entry_param_count)
+        << Attr->getStage() << FD->getNumParams()
+        << /*Special message for callable.*/ 3;
+  else {
+    ParmVarDecl *Param = FD->getParamDecl(0);
+    if (!(Param->getAttr<HLSLInOutAttr>() ||
+          (Param->getAttr<HLSLOutAttr>() && Param->getAttr<HLSLInAttr>())))
+      S.Diag(Param->getLocation(), diag::err_payload_requires_inout)
+          << /*payload|callable*/ 1 << Param;
+    QualType Ty = Param->getType().getNonReferenceType();
 
-  default:
-    if (!FD->getReturnType()->isVoidType())
-      S.Diag(FD->getLocation(), diag::err_raytracing_must_return_void);
+    if (!(hlsl::IsHLSLCopyableAnnotatableRecord(Ty)))
+      S.Diag(Param->getLocation(), diag::err_payload_attrs_must_be_udt)
+          << /*payload|attributes|callable*/ 2 << Param;
   }
+  return;
+}
 
-  switch (Stage) {
-  case DXIL::ShaderKind::Callable: {
-    if (FD->getNumParams() != 1)
-      S.Diag(FD->getLocation(), diag::err_raytracing_entry_param_count)
-          << Attr->getStage() << FD->getNumParams()
-          << /*Special message for callable.*/ 3;
-    else {
-      ParmVarDecl *Param = FD->getParamDecl(0);
-      if (!(Param->getAttr<HLSLInOutAttr>() ||
-            (Param->getAttr<HLSLOutAttr>() && Param->getAttr<HLSLInAttr>())))
-        S.Diag(Param->getLocation(), diag::err_payload_requires_inout)
-            << /*payload|callable*/ 1 << Param;
-      QualType Ty = Param->getType().getNonReferenceType();
-
-      if (!(hlsl::IsHLSLCopyableAnnotatableRecord(Ty)))
-        S.Diag(Param->getLocation(), diag::err_payload_attrs_must_be_udt)
-            << /*payload|attributes|callable*/ 2 << Param;
-    }
+void DiagnoseMissOrAnyHitEntry(Sema &S, FunctionDecl *FD, HLSLShaderAttr *Attr,
+                               bool isMiss) {
```

Comment:
Rather than a bool here, why don't we just pass in the enum value? I feel like that will make the code easier to understand.